heterocl/ast/ir_builder.py

```diff
@@ -1406,8 +1408,59 @@ def build_bit_reverse_op(self, op: ast.BitReverseOp, ip):
     def build_constant_tensor_op(self, op: ast.ConstantTensorOp, ip):
         loc = Location.file(op.loc.filename, op.loc.lineno, 0)
         dtype = hcl_dtype_to_mlir(op.dtype, signless=True)
-        val = op.values
-        value_attr = DenseElementsAttr.get(val)
+        shape = op.values.shape
+        if isinstance(op.dtype, (htypes.Int, htypes.UInt)):
+            # The following code has several steps to convert the numpy array to have
+            # the correct data type in order to create an MLIR constant tensor.
+            # Since MLIR-NumPy Python interface only supports byte-addressable data types,
+            # we need to change the data type of the array to have the minimum number of bytes
+            # that can represent the target bitwidth.
+            # e.g., hcl.const_tensor(arr, dtype=hcl.Int(20)) (6*6 array)
+            #       which requires 20 bits (3 bytes) to represent each element
+            # declaration: 6*6*i20
+            # numpy input: 6*6*i64
+            # 1. Decompose the original i32 or i64 array into a structured array of uint8
+            #  -> decompose: 6*6*8*i8
+            if op.dtype.bits == 1:
+                val = op.values
+                array = np.packbits(val, axis=None, bitorder="little")
+                value_attr = DenseElementsAttr.get(array, shape=val.shape, type=dtype)
+            else:
+                decomposed_np_dtype = np.dtype(
+                    (
+                        op.values.dtype,
+                        {
+                            f"f{i}": (np.uint8, i)
+                            for i in range(op.values.dtype.itemsize)
+                        },
+                    )
+                )
+                val = op.values.view(decomposed_np_dtype)
+                # 2. Compose the uint8 array into a structured array of target bitwidth
+                # This is done by taking the first several bytes of the uint8 array
+                n_bytes = int(np.ceil(dtype.width / 8))
+                new_dtype = np.dtype(
+                    {
+                        "names": [f"f{i}" for i in range(n_bytes)],
+                        "formats": ["u1"] * (n_bytes - 1)
+                        + (["i1"] if isinstance(dtype, htypes.Int) else ["u1"]),
```

Comment:
Another question, why does the last byte need to have sign information? I guess the last byte is the most significant byte, but I'm not sure why its format is `i1` when it's signed

Reply:
I fixed this one, but actually the signedness here does not matter a lot, since the data type of the final global tensor is determined by the user-defined type (several lines below)